components/integrations_marketplace/marketplace_modal.jsx

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {FormattedMessage} from 'react-intl';
+import {Tabs, Tab} from 'react-bootstrap';
+
+import FullScreenModal from 'components/widgets/modals/full_screen_modal';
+import RootPortal from 'components/root_portal';
+import QuickInput from 'components/quick_input';
+import LocalizedInput from 'components/localized_input/localized_input';
+import PluginIcon from 'components/svg/plugin_icon.jsx';
+import LoadingScreen from 'components/loading_screen.jsx';
+
+import {t} from 'utils/i18n';
+import {localizeMessage} from 'utils/utils';
+
+import MarketplaceItem from './marketplace_item';
+
+import './marketplace_modal.scss';
+
+const MarketplaceTabs = {
+    ALL_PLUGINS: 'allPlugins',
+    INSTALLED_PLUGINS: 'installed',
+};
+
+export default class MarketplaceModal extends React.Component {
+    static propTypes = {
+        show: PropTypes.bool,
+        installedPlugins: PropTypes.array.isRequired,
+        marketplacePlugins: PropTypes.array.isRequired,
+        actions: PropTypes.shape({
+            closeModal: PropTypes.func.isRequired,
+            getMarketplacePlugins: PropTypes.func.isRequired,
+        }).isRequired,
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            tabKey: MarketplaceTabs.ALL_PLUGINS,
+            loading: true,
+        };
+    }
+
+    componentWillMount() {
+        this.getMarketplacePlugins();
+    }
+
+    getMarketplacePlugins = async () => {
+        await this.props.actions.getMarketplacePlugins();
+
+        this.setState({loading: false});
+    }
+
+    handleSelect = (key) => {
+        this.setState({tabKey: key});
+    }
+
+    close = () => {
+        this.props.actions.closeModal();
+    }
+
+    changeTab = (tabKey) => {
+        this.setState({tabKey});
+    }
+
+    getPluginsListContent = (pluginsArray, installedList) => {
+        let noPluginsMessage = (
+            <FormattedMessage
+                id='marketplace_modal.no_plugins'
+                defaultMessage='There are no plugins available at this time.'
+            />);
+        if (installedList) {
+            noPluginsMessage = (
+                <FormattedMessage
+                    id='marketplace_modal.no_plugins_installed'
+                    defaultMessage='You do not have any plugins installed.'
+                />);
+        }
+
+        return (<div className='more-modal__list'>
```

Comment:
I think we can simplify it with a small bit of repetition. I would replace the whole block with something like:

```js
if (pluginsArray.length === 0) {
    let noPluginsMessage = ...
    return <div...
}
return <div...>{pluginsArray.map(p) => {...}}</div>
```

This block is a bit hard to follow and with a bit extra structure can be more simple.

Reply:
Done. Much better! Thanks